### PR TITLE
Replace webpack with Vite

### LINKED ISSUE #1804: Maintenance: move from webpack to vite
As opposed to earlier plans, I decided to move to Vite ("Veet", https://vitejs.dev/), instead of tsup/esbuild, just due to better developer experience, better plugin support and almost the same performance.

This will be done once no webpack config or reference to it is present anymore, but instead we have a Vite config which will take over.

### PR DESCRIPTION
**Pull request by chmanie**
## Description

I'm _really_ sorry this got so big again but I wanted to provide you with the best possible developer experience! Hope you can forgive me.

**Breaking changes**

* Vite uses the index.html as the entrypoint. I moved it to the root for the least hassle (Vite expects it there)
* To start the dev server / build the production bundle (formerly known as `npm run webpack[:prod]`) is now `npm run frontend` and `npm run build` respectively to be agnostic of the build system in the future for these commands
* `.tsx` files are now only allowed _one_ export which must be a component
* `process.env` is now `import.meta.env` - Furthermore we have access to `import.meta.env.(DEV|PROD)` and [other env vars](https://vitejs.dev/guide/env-and-mode#env-variables)
* New `.env` regime!! 
 
### **Please read this!**

The .env variables are now set in different files. We have an `.env.example` file which defines all available variables and gives descriptions and example values for it. Then there's a `.env.local` which defines all these variables for the local development environment. Furthermore there is `.env.development` which contains the environment variables that do not exist on production. Types for all the production env variables are generated automatically. There is NO `.env` anymore! Please delete it locally!!

tl;dr:
- Delete your local `.env` file
- Variables that are needed in production need to go into `.env.example` and `.env.local`

**New stuff** ✨

* Vite is now in charge of creating the frontend builds. It's _fast_ :zap: and reduced the bundle size by more than 25%
* Faster production builds (more than 3x faster!)

**Changes** 🏗

* Use Vite for storybook as well

**Deletions** ⚰️

* Webpack :yawning_face:
* Unused modules and css cleanup
* I removed all svgo plugins as we do not need them anymore (we barely have custom svgs in there after the move to phosphor)
* Removed the different tsconfig and webpack configs
* Removed loads of dev dependencies

Resolves #1804

### CODE REVIEW

**Review by iamsamgibbs — COMMENTED** [review]
Great work on this! Looks like it was another big undertaking! Builds and all looking as expected for the most part.

I've noticed I'm not getting the favicon.

Screenshot 2024-02-19 at 12 17 11

And I'm also not able to create any actions.

Screenshot 2024-02-19 at 12 25 11

**Comment by chmanie** [comment]
@iamsamgibbs thank you, I'll take a look!

**Comment by chmanie** [comment]
@iamsamgibbs Should be good now, thanks again for spotting those!

**Review by ArmandoGraterol — CHANGES_REQUESTED** [review]
Good work with all of this! It's compiling lightning fast for me ⚡ 

I did notice 2 things that might be related to this PR as I didn't find those issues on `master`:

- The notification that appears when you are installing/enabling an extension is now a bit broken (icons):
Screenshot 2024-02-19 at 13 40 23

- The transaction descriptions are not being formatted in the gas station:
Screenshot 2024-02-19 at 13 40 38

**Review by jakubcolony — DISMISSED** [review]
As thorough as you can be with PRs of this size - nothing sticks out apart from aforementioned. And I believe there's gonna be some kind of QA before it goes live anyway?

The speed improvements are huge, great job putting all of this together! 🎉 

Can we keep the port as `9091`? Apart from being used to it, there seems to be 4 files still referencing `localhost:9091` so they'd need updating either way.

**Review by iamsamgibbs — APPROVED** [review]
Other than the issues Armando noted, this is working great. Thanks for resolving the previous issues.

**Comment by chmanie** [comment]
OK, I think I fixed everything.

@ArmandoGraterol's Toast style and debug messages and @jakubcolony's port change

**Review by ArmandoGraterol — DISMISSED** [review]
Thank you for the changes Chris 🙏

**Review by iamsamgibbs — APPROVED** [review]

**Comment by rdig** [comment]
@chmanie can you please hold off merging this until later today, as I want to test it locally as well, and just couldn't find the time yesterday

File: .env.example [hunk 1]
```diff
@@ -1,40 +1,46 @@
...
 # Needed for currency conversion. For testing, see: https://support.coingecko.com/hc/en-us/articles/21880397454233
-COINGECKO_API_KEY=<coingecko-api-key>
+VITE_COINGECKO_API_KEY=
+
+# For debugging
+VITE_DEBUG=true
+
+# @NOTE Only on production
+# It became apparent that we need a "live" way to check what commit we're running on while debugging issues strait into production
+VITE_PROD_COMMIT_HASH=
```
Comment (rdig) [inline .env.example:45]: This will need to be injected automatically somehow. Either by vite or by Docker, but it's really not an option to add this manually each time

File: flake.nix [hunk 2]
```diff
@@ -2,14 +2,15 @@
   description = "Flake to develop the CDapp using nix(OS)";
```
Comment (rdig) [inline flake.nix:2]: What does this file / util actually do ?

File: src/components/common/ColonyActionsTable/ColonyActionsTable.tsx [hunk 3]
```diff
@@ -4,40 +4,190 @@ import {
...
 
+const MSG = defineMessages({
+  tableHeaderLatestActivity: {
+    id: 'common.ColonyActionsTable',
+    defaultMessage: 'Latest activity',
+  },
+});
+
+// @TODO: not-a-hook please use component instead
+const useRenderSubComponent = (
```
[the remark this answers is not in the document]
Comment (rdig) [reply src/components/common/ColonyActionsTable/ColonyActionsTable.tsx:59]: Oh never mind. I get it now. You've extracted the actual components from the hooks, so that they're now proper hooks and not hybrids.

Nice work!

File: vite.config.mts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import react from '@vitejs/plugin-react-swc';
```
Comment (rdig) [inline vite.config.mts:1]: Is there a separate config file for the vite production build ?

**Review by rdig — CHANGES_REQUESTED** [review]
Changes seems sensible, however I can't get it to work on my end

1. Can start it fine
Screenshot from 2024-02-21 19-18-30

2. The page seems to try to load and the network tab shows files being brought over
Screenshot from 2024-02-21 19-12-44
Screenshot from 2024-02-21 19-12-50
And after a while it just dies:
Screenshot from 2024-02-21 19-14-11

Here's how it looks in action:

[vite-not-rendering.webm](https://github.com/JoinColony/colonyCDapp/assets/1193222/432fdae1-0ea5-4666-b02c-8d4fadc9bfd1)

For reference, I've nuked `node_modules` and rebuild everything from the ground up two times.

File: .env.development [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Environment variables for the frontend in development mode
```
[the remark this answers is not in the document]
Comment (chmanie) [reply .env.development:1]: Good point, forgot about that one here!

File: .env.example (hunk 1 above)
Comment (chmanie) [reply .env.example:45]: My intention was for this file to serve as an example for all the environment variables that are available. It is indeed being set in the Dockerfile.
If you want, I can remove it here.

File: flake.nix (hunk 2 above)
Comment (chmanie) [reply flake.nix:2]: This one is like an nvmrc for nixOS. If you don't want it here, I can remove it.

File: vite.config.mts (hunk 4 above)
Comment (chmanie) [reply vite.config.mts:1]: There is not, currently as it isn't necessary. Do you have anything that needs changing in the production build?

File: src/components/common/ColonyActionsTable/ColonyActionsTable.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (chmanie) [reply src/components/common/ColonyActionsTable/ColonyActionsTable.tsx:59]: It's still a TODO though as I think these should be individual components in own files and not hooks.

**Comment by chmanie** [comment]
> Changes seems sensible, however I can't get it to work on my end

For some reason I'm also having trouble opening it in Thorium. Chromium (121) and Firefox work perfectly fine though. Which browser are you using?

**Comment by rdig** [comment]
> > Changes seems sensible, however I can't get it to work on my end
> 
> For some reason I'm also having trouble opening it in Thorium. Chromium (121) and Firefox work perfectly fine though. Which browser are you using?

Just plain Chrome, v121

Screenshot from 2024-02-22 19-01-03

File: vite.config.mts (hunk 4 above)
Comment (rdig) [reply vite.config.mts:1]: Not really. I was just asking as I was trying to draw parallels to webpack.

File: flake.nix (hunk 2 above)
Comment (rdig) [reply flake.nix:2]: No, leave it in. I was just unsure of what it does.

File: .env.example (hunk 1 above)
Comment (rdig) [reply .env.example:45]: Fine then. As long as I't automatic I have no issues.

File: .env.development (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdig) [reply .env.development:1]: Sure, but we use `URL` in code as well to extract it's various origin/hosts/protocols and use it directly in the app

https://github.com/JoinColony/colonyCDapp/blob/master/src/constants/index.ts#L296

[the remark this answers is not in the document]
Comment (chmanie) [reply .env.development:1]: So do you want me to add it here even though it's not necessary?

[the remark this answers is not in the document]
Comment (rdig) [reply .env.development:1]: In the above example we use it for display purposes mostly throughout the app, so unless you refactored that out, it needs to be added in, so we can reference it

**Review by rdig — APPROVED** [review]
Very nice getting to the bottom of this. It now works fine on my machine as well, regardless of browser.

Just one quick note that I've used:
```bash
ulimit -b 2048
```
rather then your instructions, which I think it's should be preferred, since it just modifies the user limits, and does not involve any file changes, so it's quicker _(we can even create a script for it)_

Here's a demo of the vite server working:
[vite-working.webm](https://github.com/JoinColony/colonyCDapp/assets/1193222/0003a89e-17dd-4dbe-92ab-131d3b52b55f)

And here's a recording of the network tab, which, just looking at, shows why we ran into this problem in the first place:
[vite-network.webm](https://github.com/JoinColony/colonyCDapp/assets/1193222/3c74fe4d-19b3-4ea6-be06-35fc732c4078)

So this is all good from my end, so let's just hope @bogdan-1337 comes up with a solution for the variables then we can finally merge this!!!

**Review by ArmandoGraterol — APPROVED** [review]
Still working as expected for me 👍

**Review by jakubcolony — DISMISSED** [review]
It runs, I tested HMR, `build`, storybook, haven't spotted any obvious regressions.

Great work @chmanie, this is so much faster and nicer experience.

Just a question around the `.env.local` file, should that be committed to the repo?

**Comment by chmanie** [comment]
> Just a question around the .env.local file, should that be committed to the repo?

Yes, as these are values that are only used in development mode. There should be no secret keys exposed here. If we at some point do have secrets that are needed throughout development I can add a solution for that, too.

**Review by jakubcolony — APPROVED** [review]

**Review by rdig — DISMISSED** [review]
Managed to get around and do another test on this. Everything works as expected, even with the `ulimit` file descriptors reset back to my original `1024`, which is very weird _(I've removed your `ulimit` call from the npm `frontend` script)_

Screenshot from 2024-03-14 21-13-25

So seeing as everything works, I see no reason why we should delay  merging this further.

Very nice job Chris! :100:

**Comment by chmanie** [comment]
I needed to rebase, @rdig, @jakubcolony and/or @ArmandoGraterol would you like to help me out with a review again?

**Review by rdig — APPROVED** [review]

**Comment by rdig** [comment]
Approved it, but maybe wait until Monday to merge, in case anyone wants to work over the weekend and runs into trouble rebasing

**Review by ArmandoGraterol — APPROVED** [review]
Still working great for me 💯 Excellent stuff Chris!

**Review by jakubcolony — APPROVED** [review]
